### ISSUE TITLE
haproxy=2.0.7: A bogus APPCTX is spinning and refuses to die

### ISSUE THREAD

**Issue by max2k1** [haproxy/haproxy#277]
Sometimes on reload of 2.0.5 I got this in logs:
```
A bogus APPCTX [0x7fc1a06ff0e0] is spinning at 122591 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x557eb7f4e630 src=xxx fe=yyy be=yyy dst=<PEER> rqf=c48202 rqa=0 rpf=80048202 rpa=0 sif=EST,200040 sib=EST,244058 af=(nil),0 csf=0x7fc1ac085680,200 ab=0x7fc1a06ff0e0,7 csb=(nil),0 cof=0x557eb7fe9f50,201366:PASS(0x7fc16c0ad080)/RAW((nil))/tcpv6(1275) cob=(nil),0:NONE((nil))/NONE((nil))/NONE(0) ]
```

...and this in coredump:
```
Program terminated with signal SIGABRT, Aborted.
#0  0x00007fc1e7c0b428 in raise () from /lib/x86_64-linux-gnu/libc.so.6
[Current thread is 1 (Thread 0x7fc1e95021c0 (LWP 3539))]
(gdb) bt
#0  0x00007fc1e7c0b428 in raise () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007fc1e7c0d02a in abort () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x0000557ea75b8b5e in stream_dump_and_crash (obj=obj@entry=0x7fc1a06ff0e0, rate=122591) at src/stream.c:2983
#3  0x0000557ea7696d2d in task_run_applet (t=0x7fc1746a5790, context=0x7fc1a06ff0e0, state=<optimized out>) at src/applet.c:80
#4  0x0000557ea7692fd5 in process_runnable_tasks () at src/task.c:414
#5  0x0000557ea75fbc18 in run_poll_loop () at src/haproxy.c:2517
#6  run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:2638
#7  0x0000557ea7558cb7 in main (argc=<optimized out>, argv=0x7ffd7db57428) at src/haproxy.c:3315
```
It happens only on heavy-loaded haproxy-instances with nearly 2000 backends with http-checks enabled and sticky-tables usage. 

```
Version: haproxy -vv
HA-Proxy version 2.0.5-1 2019/08/27 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_REGPARM=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_TFO=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=56).
Built with OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
Running on OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.3.1
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.21 2016-01-12
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace
```

**Comment by lukastribus** [comment]
Please upgrade to `2.0.6` first of all.

**Comment by max2k1** [comment]
> Please upgrade to `2.0.6` first of all.

I can't until https://github.com/haproxy/haproxy/issues/278 is fixed
Got same problems on my config

**Comment by lukastribus** [comment]
Can you revert the check patch on top of 2.0.6, as per:
https://github.com/haproxy/haproxy/issues/278#issuecomment-531754172

**Comment by max2k1** [comment]
> Can you revert the check patch on top of 2.0.6, as per:
> [#278 (comment)](https://github.com/haproxy/haproxy/issues/278#issuecomment-531754172)

Did it. Now need some time to check if it helps. Will write back in a few days.

**Comment by max2k1** [comment]
Didn't help:
```
2019-09-18T11:57:39+03:00 haproxy[34713] local0.emerg A bogus APPCTX [0x7f5cc807fe20] is spinning at 100159 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x7f5cc807f600 src=<PEER> fe=yyy be=yyy dst=unknown rqf=848000 rqa=0 rpf=80448000 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x7f5cc807fe20,7 csf=(nil),0 ab=(nil),0 csb=0x7f5cc807fd50,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x7f5cc807fba0,202366:PASS(0x7f5cc807fda0)/RAW((nil))/tcpv6(129) ]
```

Backtrace
```
Program terminated with signal SIGABRT, Aborted.
#0  0x00007f5d53624428 in raise () from /lib/x86_64-linux-gnu/libc.so.6
[Current thread is 1 (Thread 0x7f5d24ff9700 (LWP 34744))]
(gdb) bt
#0  0x00007f5d53624428 in raise () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007f5d5362602a in abort () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x000055bd5ed5dfae in stream_dump_and_crash (obj=obj@entry=0x7f5cc807fe20, rate=100159) at src/stream.c:2983
#3  0x000055bd5ee3c64d in task_run_applet (t=0x7f5cc807ffd0, context=0x7f5cc807fe20, state=<optimized out>) at src/applet.c:80
#4  0x000055bd5ee388f5 in process_runnable_tasks () at src/task.c:414
#5  0x000055bd5eda13e8 in run_poll_loop () at src/haproxy.c:2517
#6  run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:2638
#7  0x00007f5d547266ba in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#8  0x00007f5d536f641d in clone () from /lib/x86_64-linux-gnu/libc.so.6
```

Version:
```
$ haproxy -vv
HA-Proxy version 2.0.6-2 2019/09/16 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_REGPARM=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_TFO=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=56).
Built with OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
Running on OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.3.1
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.21 2016-01-12
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace
```

**Comment by max2k1** [comment]
2.0.7 - still crashing:
```
2019-10-02T12:44:24+03:00 haproxy[16249] local0.emerg A bogus APPCTX [0x7fab2c07fe20] is spinning at 100141 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x7fab2c080150 src=<PEER> fe=xxx be=xxx dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x7fab2c07fe20,7 csf=(nil),0 ab=(nil),0 csb=0x7fab2c0808a0,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x7fab2c0806f0,202366:PASS(0x7fab2c0808f0)/RAW((nil))/tcpv6(807) ]
```

<details>
<summary>backtrace</summary>

```
Program terminated with signal SIGABRT, Aborted.
#0  0x00007fabb7d3d428 in raise () from /lib/x86_64-linux-gnu/libc.so.6
[Current thread is 1 (Thread 0x7fab7d7fa700 (LWP 16282))]
(gdb) bt
#0  0x00007fabb7d3d428 in raise () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007fabb7d3f02a in abort () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x000055a77c4ad08e in stream_dump_and_crash (obj=obj@entry=0x7fab2c07fe20, rate=100141) at src/stream.c:2983
#3  0x000055a77c58b71d in task_run_applet (t=0x7fab2c07ffd0, context=0x7fab2c07fe20, state=<optimized out>) at src/applet.c:80
#4  0x000055a77c5879c5 in process_runnable_tasks () at src/task.c:414
#5  0x000055a77c4f0468 in run_poll_loop () at src/haproxy.c:2517
#6  run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:2638
#7  0x00007fabb8e3f6ba in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#8  0x00007fabb7e0f41d in clone () from /lib/x86_64-linux-gnu/libc.so.6
```
</details>

<details>
<summary>version</summary>

```
haproxy -vv
HA-Proxy version 2.0.7-1 2019/09/30 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_REGPARM=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_TFO=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=56).
Built with OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
Running on OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.3.1
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.21 2016-01-12
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace
```
</details>

**Comment by wtarreau** [comment]
Thank you. I'm starting to understand it better. Visibly the peers applet is facing a buffer full condition on the output path and because of this cannot process the input data (typically because an ACK would need to be emitted). But apparently it's leaving indicating that it's waiting for more data so it's immediately called again. Without the loop protection what would happen would be random CPU usage spikes when this buffer is full, a situation that you might possibly have witnessed with older versions. I'm having a look at this now.

**Comment by wtarreau** [comment]
I don't see anything wrong in the peers applet. Doing some quick math, at 100k calls per second for 16kB buffers, that's 1.6 GB of traffic exchanged in one second, or 12 Gbps of peers traffic. Are you running with huge stick tables ? Maybe we're catching a false positive here and we'd need to include an extra check to figure if the applet managed to return something or not.

**Comment by max2k1** [comment]
We have about 20 stick-tables per-instance config synchronized within a group of 3 servers. All tables are of same kind. Typical usage: calculate current request-rate across a cluster. Example:

<details><summary>server1</summary>

```
peers slb_local
  peer server1
  peer server2
  peer server3

frontend fe
  http-request track-sc0 always_true table backend-server1
  http-request set-var(req.rr1) sc0_http_req_rate(backend-server1)
  http-request set-var(req.rr2) sc0_http_req_rate(backend-server2)
  http-request set-var(req.rr3) sc0_http_req_rate(backend-server3)
  http-request set-var(req.rr0) var(req.rr1),add(req.rr2),add(req.rr3),div(10)

backend backend-server1
  stick-table type integer size 1 expire 30s peers slb_local store http_req_rate(10s),http_err_rate(10s)

backend backend-server2
  stick-table type integer size 1 expire 30s peers slb_local store http_req_rate(10s),http_err_rate(10s)

backend backend-server3
  stick-table type integer size 1 expire 30s peers slb_local store http_req_rate(10s),http_err_rate(10s)
```
</details>

<details><summary>server2</summary>

```
peers slb_local
  peer server1
  peer server2
  peer server3

frontend fe
  http-request track-sc0 always_true table backend-server2
  http-request set-var(req.rr1) sc0_http_req_rate(backend-server1)
  http-request set-var(req.rr2) sc0_http_req_rate(backend-server2)
  http-request set-var(req.rr3) sc0_http_req_rate(backend-server3)
  http-request set-var(req.rr0) var(req.rr1),add(req.rr2),add(req.rr3),div(10)

backend backend-server1
  stick-table type integer size 1 expire 30s peers slb_local store http_req_rate(10s),http_err_rate(10s)

backend backend-server2
  stick-table type integer size 1 expire 30s peers slb_local store http_req_rate(10s),http_err_rate(10s)

backend backend-server3
  stick-table type integer size 1 expire 30s peers slb_local store http_req_rate(10s),http_err_rate(10s)
```
</details>

<details><summary>server3</summary>

```
peers slb_local
  peer server1
  peer server2
  peer server3

frontend fe
  http-request track-sc0 always_true table backend-server3
  http-request set-var(req.rr1) sc0_http_req_rate(backend-server1)
  http-request set-var(req.rr2) sc0_http_req_rate(backend-server2)
  http-request set-var(req.rr3) sc0_http_req_rate(backend-server3)
  http-request set-var(req.rr0) var(req.rr1),add(req.rr2),add(req.rr3),div(10)

backend backend-server1
  stick-table type integer size 1 expire 30s peers slb_local store http_req_rate(10s),http_err_rate(10s)

backend backend-server2
  stick-table type integer size 1 expire 30s peers slb_local store http_req_rate(10s),http_err_rate(10s)

backend backend-server3
  stick-table type integer size 1 expire 30s peers slb_local store http_req_rate(10s),http_err_rate(10s)
```
</details>

As you can see, tables are tiny but synchronization packet rate is high. Speaking of workload bandwidth, it is about from 1 upto 2Gbit per haproxy instance.

**Comment by wtarreau** [comment]
So if the tables are tiny there is no way you can end up with full buffers. We don't accumulate multiple messages for a same key in an announce, even if you have many entries for the same host, only the most recent one gets sent. The fact that the issue happens on reload is what made me think you could have a huge table with millions of keys and lots of data (e.g. keeping counters per user agent over a week or lots of stats per IP address). I'll need to have a deeper look into the learning part of the code that is used during the reload, just in case there's anything specific there.

**Comment by wtarreau** [comment]
Thinking a bit more about it we could imagine that if you're dealing with long lasting connections and your requests are evenly spread over the 2000 backends, and these backends each use their own tables, we could imagine that entries take room and are updated while being learned, resulting in a high message rate during this period. I'll try to setup a test case for this and see how far it can go. If I'm seeing it can reach orders of thousands of calls per second or more, I'll have to figure a more reliable way to detect looping applets (such as figuring which ones neither consumed nor produced data).

**Comment by wtarreau** [comment]
Well for now all I managed to do was to crash the old process in a segfault during the reload, so that's different from what you're seeing... I'm running with 50000 backends and as many stick-tables.

Program terminated with signal SIGSEGV, Segmentation fault.
#0  process_peer_sync (task=0x23bba8f0, context=0x1a34af0, state=<optimized out>) at src/peers.c:2736
2736                    if (ps->flags & PEER_F_TEACH_COMPLETE) {
[Current thread is 1 (Thread 0x7fdeecb12700 (LWP 26043))]
(gdb) p ps
$1 = (struct peer *) 0x0
(gdb) bt
#0  process_peer_sync (task=0x23bba8f0, context=0x1a34af0, state=<optimized out>) at src/peers.c:2736
#1  0x000000000054ab95 in process_runnable_tasks () at src/task.c:426
#2  0x00000000004bf93c in run_poll_loop () at src/haproxy.c:2610
#3  run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:2727
#4  0x00007fdef206719c in start_thread () from /lib64/libpthread.so.0
#5  0x00007fdef1a504af in clone () from /lib64/libc.so.6

Still investigating.

**Comment by wtarreau** [comment]
So I've met other issues whiel working on this. A huge table lookup cost that can cause pauses so long that it can trigger the watchdog when you have many tables (100k), a case where it's frequent that two connections appear in parallel during the reload, and some cases where the old process seems to stop before all elements are taught. I've tried with moderate (125) and low (3) table counts, with either few or huge amounts of entries. I could never reach numbers as high as yours, the highest wakeup rate I've seen was around 3000/s (which is already huge), so I really think that something is misbehaving somewhere. I'll try to make sure the applet wakeup rate is only applied to applets having done nothing, but we definitely have several issues in the peers synchronization part.

**Comment by max2k1** [comment]
I forget to mention operating mode of our instances. Maybe it could help somehow. For example maybe there're some issues with stick-tables and hitless reload (I just imagine).

global part of config is
```
global
  daemon
  stats socket /var/run/haproxy.sock mode 700 level admin expose-fd listeners
  master-worker
  nbthread 32
```

and we run haproxy as follows
```
/usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -sf XXXXX -x /var/run/haproxy.sock
```

**Comment by wtarreau** [comment]
Thanks, that's mainly how I was using it as well :-/

**Comment by wtarreau** [comment]
OK I think I found how to artificially trigger this check. I modified the send() code in rawsock to let data leave in small chunks (one TCP segment at a time) and requested a huge stats page made of 50000 proxies. This manages to call the applet fast enough for it to fail. But in your case only very small messages are sent and there's a single key so I hardly see how this could happen. Regardless, I want to address this risk so I'm going to add extra conditions to decide if it's an issue (i.e. applet reporting missing room while the buffer is empty, or not processing input data without asking for more).

**Comment by ingshtrom** [comment]
Is there anything else that I can help with here? I'm seeing this as well in our production environment running the official haproxy:2.0.7 Docker image.

I don't follow much of the preceding conversation, but you mentioned loading a "huge stats page". Ours is only about 500 lines long (doing a CSV export), so tiny compared to yours, but we do have [haproxy_exporter](https://github.com/prometheus/haproxy_exporter) scraping the stats page every 15 seconds. 

Are you thinking that could contribute to this or only in extreme cases like your test environment?

Here are some examples we've gotten over the last week if they help at all (CSV format):
```
"_messagetimems","_messagetime","container_name","message"
"1571678669875","10/21/2019 13:24:29.875 -0400","k8s_haproxy_haproxy-router-f5p4l_routing-external-default_13f6ebae-e6bc-11e9-ba98-0242ac11000c_83","[ALERT] 293/172429 (1075) : A bogus APPCTX [0x55b4c9344fc0] is spinning at 100579 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x55b4c93452f0 src=<PEER> fe=ip-10-1-25-2.ec2.internal be=ip-10-1-25-2.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x55b4c9344fc0,7 csf=(nil),0 ab=(nil),0 csb=0x55b4c9345a40,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x55b4c9345890,202366:PASS(0x55b4c9345a90)/RAW((nil))/tcpv4(47) ]"
"1571674702858","10/21/2019 12:18:22.858 -0400","k8s_haproxy_haproxy-router-9f2wt_routing-external-default_3c1a9963-e6bc-11e9-ba98-0242ac11000c_10","[ALERT] 293/161822 (398) : A bogus APPCTX [0x5588af717ea0] is spinning at 107163 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x5588af7181d0 src=<PEER> fe=ip-10-1-41-9.ec2.internal be=ip-10-1-41-9.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x5588af717ea0,7 csf=(nil),0 ab=(nil),0 csb=0x5588af718920,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x5588af718770,202366:PASS(0x5588af718970)/RAW((nil))/tcpv4(45) ]"
"1571673667651","10/21/2019 12:01:07.651 -0400","k8s_haproxy_haproxy-router-9cpgm_routing-external-registry_93390ca0-f40d-11e9-b1c3-0242ac110005_0","[ALERT] 293/160107 (1580) : A bogus APPCTX [0x7f3fac05d0c0] is spinning at 100009 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x7f3fac05d3f0 src=<PEER> fe=ip-10-1-8-180.ec2.internal be=ip-10-1-8-180.ec2.internal dst=unknown rqf=848000 rqa=0 rpf=80448000 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x7f3fac05d0c0,7 csf=(nil),0 ab=(nil),0 csb=0x7f3fac05db40,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x7f3fac05d990,202366:PASS(0x7f3fac05db90)/RAW((nil))/tcpv4(48) ]"
"1571673575703","10/21/2019 11:59:35.703 -0400","k8s_haproxy_haproxy-router-9f2wt_routing-external-default_3c1a9963-e6bc-11e9-ba98-0242ac11000c_9","[ALERT] 293/155935 (24178) : A bogus APPCTX [0x56390bfebf40] is spinning at 120552 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x56390bfec270 src=<PEER> fe=ip-10-1-41-9.ec2.internal be=ip-10-1-41-9.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x56390bfebf40,7 csf=(nil),0 ab=(nil),0 csb=0x56390bfec9c0,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x56390bfec810,202366:PASS(0x56390bfeca10)/RAW((nil))/tcpv4(46) ]"
"1571449574943","10/18/2019 21:46:14.943 -0400","k8s_haproxy_haproxy-router-9f2wt_routing-external-default_3c1a9963-e6bc-11e9-ba98-0242ac11000c_8","[ALERT] 291/014614 (1731) : A bogus APPCTX [0x55b156a76e50] is spinning at 102165 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x55b156a77180 src=<PEER> fe=ip-10-1-41-9.ec2.internal be=ip-10-1-41-9.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x55b156a76e50,7 csf=(nil),0 ab=(nil),0 csb=0x55b156a778d0,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x55b156a77720,202366:PASS(0x55b156a77920)/RAW((nil))/tcpv4(45) ]"
"1571442793914","10/18/2019 19:53:13.914 -0400","k8s_haproxy_haproxy-router-9f2wt_routing-external-default_3c1a9963-e6bc-11e9-ba98-0242ac11000c_7","[ALERT] 290/235313 (183) : A bogus APPCTX [0x560e7ccdfe50] is spinning at 104937 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x560e7cce0180 src=<PEER> fe=ip-10-1-41-9.ec2.internal be=ip-10-1-41-9.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,204048 sib=EST,2000d8 af=0x560e7ccdfe50,7 csf=(nil),0 ab=(nil),0 csb=0x560e7cce08d0,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x560e7cce0720,202306:PASS(0x560e7cce0920)/RAW((nil))/tcpv4(45) ]"
"1571442267822","10/18/2019 19:44:27.822 -0400","k8s_haproxy_haproxy-router-9ql2x_routing-external-default_61c1785d-e6bc-11e9-ba98-0242ac11000c_72","[ALERT] 290/234427 (2383) : A bogus APPCTX [0x5556ed718c70] is spinning at 109229 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x5556ed718fa0 src=<PEER> fe=ip-10-1-25-98.ec2.internal be=ip-10-1-25-98.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,204048 sib=EST,2000d8 af=0x5556ed718c70,7 csf=(nil),0 ab=(nil),0 csb=0x5556ed7196f0,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x5556ed719540,202306:PASS(0x5556ed719740)/RAW((nil))/tcpv4(44) ]"
"1571442260782","10/18/2019 19:44:20.782 -0400","k8s_haproxy_haproxy-router-9f2wt_routing-external-default_3c1a9963-e6bc-11e9-ba98-0242ac11000c_6","[ALERT] 290/234420 (23759) : A bogus APPCTX [0x55993672dc70] is spinning at 109527 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x55993672dfa0 src=<PEER> fe=ip-10-1-41-9.ec2.internal be=ip-10-1-41-9.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,204048 sib=EST,2000d8 af=0x55993672dc70,7 csf=(nil),0 ab=(nil),0 csb=0x55993672e6f0,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x55993672e540,202306:PASS(0x55993672e740)/RAW((nil))/tcpv4(44) ]"
"1571441633576","10/18/2019 19:33:53.576 -0400","k8s_haproxy_haproxy-router-9ql2x_routing-external-default_61c1785d-e6bc-11e9-ba98-0242ac11000c_72","[ALERT] 290/233353 (396) : A bogus APPCTX [0x7f430aa28510] is spinning at 100182 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x7f4314190930 src=10.1.14.179 fe=ip-10-1-25-98.ec2.internal be=ip-10-1-25-98.ec2.internal dst=<PEER> rqf=c48202 rqa=0 rpf=80048202 rpa=0 sif=EST,200040 sib=EST,244058 af=(nil),0 csf=0x7f430aab9b10,200 ab=0x7f430aa28510,7 csb=(nil),0 cof=0x7f430c678b10,201366:PASS(0x7f4307727cd0)/RAW((nil))/tcpv4(2765) cob=(nil),0:NONE((nil))/NONE((nil))/NONE(0) ]"
"1571433161766","10/18/2019 17:12:41.766 -0400","k8s_haproxy_haproxy-router-9ql2x_routing-external-default_61c1785d-e6bc-11e9-ba98-0242ac11000c_71","[ALERT] 290/211241 (885) : A bogus APPCTX [0x564886780080] is spinning at 118181 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x5648867803b0 src=<PEER> fe=ip-10-1-25-98.ec2.internal be=ip-10-1-25-98.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,204048 sib=EST,2000d8 af=0x564886780080,7 csf=(nil),0 ab=(nil),0 csb=0x564886780b00,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x564886780950,202306:PASS(0x564886780b50)/RAW((nil))/tcpv4(47) ]"
"1571430353908","10/18/2019 16:25:53.908 -0400","k8s_haproxy_haproxy-router-k76hk_routing-external-default_8937c64b-e6bc-11e9-ba98-0242ac11000c_64","[ALERT] 290/202553 (3056) : A bogus APPCTX [0x55cd80650180] is spinning at 100629 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x55cd806504b0 src=<PEER> fe=ip-10-1-14-179.ec2.internal be=ip-10-1-14-179.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x55cd80650180,7 csf=(nil),0 ab=(nil),0 csb=0x55cd80650c00,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x55cd80650a50,202366:PASS(0x55cd80650c50)/RAW((nil))/tcpv4(49) ]"
"1571349793941","10/17/2019 18:03:13.941 -0400","k8s_haproxy_haproxy-router-9f2wt_routing-external-default_3c1a9963-e6bc-11e9-ba98-0242ac11000c_5","[ALERT] 289/220313 (2308) : A bogus APPCTX [0x55e96f537e40] is spinning at 101851 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x55e96f538170 src=<PEER> fe=ip-10-1-41-9.ec2.internal be=ip-10-1-41-9.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x55e96f537e40,7 csf=(nil),0 ab=(nil),0 csb=0x55e96f5388c0,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x55e96f538710,202366:PASS(0x55e96f538910)/RAW((nil))/tcpv4(46) ]"
"1571341154742","10/17/2019 15:39:14.742 -0400","k8s_haproxy_haproxy-router-f5p4l_routing-external-default_13f6ebae-e6bc-11e9-ba98-0242ac11000c_81","[ALERT] 289/193914 (6077) : A bogus APPCTX [0x5616c217f3b0] is spinning at 115793 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x5616c217f610 src=<PEER> fe=ip-10-1-25-2.ec2.internal be=ip-10-1-25-2.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80458202 rpa=0 sif=EST,204048 sib=EST,2000d8 af=0x5616c217f3b0,7 csf=(nil),0 ab=(nil),0 csb=0x5616c217fd60,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x5616c217fbb0,202306:PASS(0x5616c217fdb0)/RAW((nil))/tcpv4(36) ]"
"1571341134844","10/17/2019 15:38:54.844 -0400","k8s_haproxy_haproxy-router-9f2wt_routing-external-default_3c1a9963-e6bc-11e9-ba98-0242ac11000c_4","[ALERT] 289/193854 (22648) : A bogus APPCTX [0x7f87a40694b0] is spinning at 100060 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x7f87a4066d40 src=10.1.14.179 fe=ip-10-1-41-9.ec2.internal be=ip-10-1-41-9.ec2.internal dst=<PEER> rqf=c48202 rqa=0 rpf=80048202 rpa=0 sif=EST,200048 sib=EST,204058 af=(nil),0 csf=0x7f87a4069dc0,200 ab=0x7f87a40694b0,7 csb=(nil),0 cof=0x7f87a4068950,201306:PASS(0x7f87a40687d0)/RAW((nil))/tcpv4(82) cob=(nil),0:NONE((nil))/NONE((nil))/NONE(0) ]"
"1571316862871","10/17/2019 08:54:22.871 -0400","k8s_haproxy_haproxy-router-f5p4l_routing-external-default_13f6ebae-e6bc-11e9-ba98-0242ac11000c_80","[ALERT] 289/125422 (12610) : A bogus APPCTX [0x7f229405ce70] is spinning at 103705 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x7f22980666e0 src=10.1.14.179 fe=ip-10-1-25-2.ec2.internal be=ip-10-1-25-2.ec2.internal dst=<PEER> rqf=c48202 rqa=0 rpf=80048202 rpa=0 sif=EST,200040 sib=EST,244058 af=(nil),0 csf=0x7f2298066690,200 ab=0x7f229405ce70,7 csb=(nil),0 cof=0x7f2294058a90,201366:PASS(0x7f2298066660)/RAW((nil))/tcpv4(64) cob=(nil),0:NONE((nil))/NONE((nil))/NONE(0) ]"
"1571302657928","10/17/2019 04:57:37.928 -0400","k8s_haproxy_haproxy-router-9ql2x_routing-external-default_61c1785d-e6bc-11e9-ba98-0242ac11000c_70","[ALERT] 289/085737 (11697) : A bogus APPCTX [0x7f044805e730] is spinning at 100549 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x7f044805ea60 src=<PEER> fe=ip-10-1-25-98.ec2.internal be=ip-10-1-25-98.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x7f044805e730,7 csf=(nil),0 ab=(nil),0 csb=0x7f044805f1b0,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x7f044805f000,202366:PASS(0x7f044805f200)/RAW((nil))/tcpv4(58) ]"
"1571291800861","10/17/2019 01:56:40.861 -0400","k8s_haproxy_haproxy-router-k76hk_routing-external-default_8937c64b-e6bc-11e9-ba98-0242ac11000c_63","[ALERT] 289/055640 (6194) : A bogus APPCTX [0x55e2e59a7e40] is spinning at 105240 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x55e2e59a8170 src=<PEER> fe=ip-10-1-14-179.ec2.internal be=ip-10-1-14-179.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x55e2e59a7e40,7 csf=(nil),0 ab=(nil),0 csb=0x55e2e59a88c0,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x55e2e59a8710,202366:PASS(0x55e2e59a8910)/RAW((nil))/tcpv4(55) ]"
"1571268895743","10/16/2019 19:34:55.743 -0400","k8s_haproxy_haproxy-router-f5p4l_routing-external-default_13f6ebae-e6bc-11e9-ba98-0242ac11000c_80","[ALERT] 288/233455 (245) : A bogus APPCTX [0x7f13e405e730] is spinning at 104943 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x7f13e405ea60 src=<PEER> fe=ip-10-1-25-2.ec2.internal be=ip-10-1-25-2.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x7f13e405e730,7 csf=(nil),0 ab=(nil),0 csb=0x7f13e405f1b0,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x7f13e405f000,202366:PASS(0x7f13e405f200)/RAW((nil))/tcpv4(46) ]"
"1571268197970","10/16/2019 19:23:17.970 -0400","k8s_haproxy_haproxy-router-f5p4l_routing-external-default_13f6ebae-e6bc-11e9-ba98-0242ac11000c_79","[ALERT] 288/232317 (5201) : A bogus APPCTX [0x55cc40110090] is spinning at 100453 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x55cc40066000 src=10.1.9.44 fe=ip-10-1-25-2.ec2.internal be=ip-10-1-25-2.ec2.internal dst=<PEER> rqf=c48202 rqa=0 rpf=80048202 rpa=0 sif=EST,200048 sib=EST,204058 af=(nil),0 csf=0x55cc40065fb0,200 ab=0x55cc40110090,7 csb=(nil),0 cof=0x55cc40065cb0,201306:PASS(0x55cc40065f30)/RAW((nil))/tcpv4(131) cob=(nil),0:NONE((nil))/NONE((nil))/NONE(0) ]"
"1571268131936","10/16/2019 19:22:11.936 -0400","k8s_haproxy_haproxy-router-k76hk_routing-external-default_8937c64b-e6bc-11e9-ba98-0242ac11000c_62","[ALERT] 288/232211 (3722) : A bogus APPCTX [0x7f577c05e730] is spinning at 101906 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x7f577c05ea60 src=<PEER> fe=ip-10-1-14-179.ec2.internal be=ip-10-1-14-179.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,204048 sib=EST,2000d8 af=0x7f577c05e730,7 csf=(nil),0 ab=(nil),0 csb=0x7f577c05f1b0,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x7f577c05f000,202306:PASS(0x7f577c05f200)/RAW((nil))/tcpv4(56) ]"
"1571258857756","10/16/2019 16:47:37.756 -0400","k8s_haproxy_haproxy-router-9ql2x_routing-external-default_61c1785d-e6bc-11e9-ba98-0242ac11000c_69","[ALERT] 288/204737 (28141) : A bogus APPCTX [0x7f808c05fd80] is spinning at 100123 calls per second and refuses to die, aborting now! Please report this error to developers [strm=0x7f808c0600b0 src=<PEER> fe=ip-10-1-25-98.ec2.internal be=ip-10-1-25-98.ec2.internal dst=unknown rqf=848202 rqa=0 rpf=80448202 rpa=0 sif=EST,244048 sib=EST,2000d0 af=0x7f808c05fd80,7 csf=(nil),0 ab=(nil),0 csb=0x7f808c060800,200 cof=(nil),0:NONE((nil))/NONE((nil))/NONE(0) cob=0x7f808c060650,202366:PASS(0x7f808c060850)/RAW((nil))/tcpv4(46) ]"
```

**Comment by wtarreau** [comment]
In your case it's the peers, you're facing the exact same issue as max2k1 that I couldn't reproduce to such numbers. It was later addressed however and the fix is already in the 2.0 tree, so 2.0.8 will be fixed. If you want to try with the latest snapshot, it should be OK.

**Comment by zstyblik** [comment]
Please, is there any ETA when 2.0.8 will be released?

**EDIT**: also, is there any way how to prevent this from happening?

**Comment by wtarreau** [comment]
Yes, it was released yesterday : https://www.mail-archive.com/haproxy@formilux.org/msg35168.html
Besides updating to 2.0.8, I don't see a way to prevent this from happening unfortunately.
Edit: sorry, it was this morning, early enough for me to believe it was yesterday :-)

**Comment by zstyblik** [comment]
@wtarreau, thank you for heads up. I must apologize as well since I've mistaken referenced commit with PR, no idea why. Also, I've checked releases at GH, but I didn't see 2.0.8 release there.

Anyway, eagerly awaiting new Debian package.

Once more, thank you :)

**Comment by wtarreau** [comment]
The debian package was already done as well, just have a look here to see how to update it: https://haproxy.debian.net/
In order to quickly check the build statuses for various debian/ubuntu versions I usually look here: https://launchpad.net/~vbernat/

**Comment by zstyblik** [comment]
Yep, that's the one. Still getting `pool/main/h/haproxy/haproxy_2.0.7-1~bpo9+1_amd64.deb` for Stretch(9). I'll wait, no worries ;)

**Comment by max2k1** [comment]
2.0.8 seems to be a solution. No more cores till now.

**Comment by capflam** [comment]
Thanks for your feedback. I'm closing the issue now.
